Add Minio to the app install command

burtonr opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This adds the ability to install Minio using the helm chart
There are flags added to allow the user to set the access key
and secret key as well as change the mode of installation

There is a safety net check to ensure that both access and
secret key are included if one is present

Signed-off-by: Burton Rheutan <rheutan7@gmail.com>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [X] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))
Resolves #56 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Installed Minio app on local microk8s cluster

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [X] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

alexellis commented:
Hi @burtonr thanks for this PR.

Since k3sup is about ease of use and working fast, I'd like to see the two credentials generated in the same way we do for OpenFaaS.

If the user provides an override for either, then don't generate.

burtonr commented:
@alexellis Minio will generate those secrets automatically by default. However, after a bit of testing I found that the values generated by Minio were always the same...
So, I've updated the installation to generate the access and secret key if they are not provided. I've also updated the final output to describe how to access those secrets as well as port-forwarding to allow access.
Let me know if you'd like anything else

alexellis commented:
@Waterdrips please can you review and try this out?

@burtonr thank you for updating, it needs a rebase now

Waterdrips commented on pkg/cmd/minio_app.go:
```diff
@@ -0,0 +1,137 @@
+		secretKey, _ := minio.Flags().GetString("secret-key")
+
+		if len(accessKey) == 0 {
+			fmt.Printf("Access Key not provided, one will be generated for you")
```
no newline at the end of these prints - the lines get joined in output 

`Access Key not provided, one will be generated for youSecret Key not provided, one will be generated for youexec:  /Users/alistairhey/.k3sup/bin/helm template minio --name minio --namespace minio --output-dir /var/folders/y2/67nr0tqd10q35gdq84_mkz2h0000gn/T/charts/minio/rendered --values /var/folders/y2/67nr0tqd10q35gdq84_mkz2h0000gn/T/charts/minio/values.yaml  --set accessKey=#### --set secretKey=######
`

burtonr commented:
Thanks for the feedback @Waterdrips :100: I'll get those things taken care of right away and be sure to test, and re-test, on a new clean cluster.

alexellis commented on pkg/cmd/minio_app.go:
```diff
@@ -0,0 +1,137 @@
+		overrides["accessKey"] = accessKey
+		overrides["secretKey"] = secretKey
+
+		if dist, _ := minio.Flags().GetBool("distributed-mode"); dist {
```
Since the flags are modes, let's just use `distributed` instead?

burtonr replied:
Good point, updated to remove the "-mode" from the flag

burtonr commented:
@Waterdrips Please could you try this again on your side?

I've taken the time to carefully clean my system (restart machine), clean the cluster (`microk8s.reset --remove-storage`), clean my local files (`rm -rf /tmp/charts`) and clean my environment (new terminal window). I've since run this about 5 times successfully. 

Please do let me know if there is still an issue. It is possible I am missing something on my side between tests that is allowing it to pass for me.

blaise-sumo commented:
These are such thorough and constructive reviews that I find myself wanting
to learn golang!

On Sun, Dec 8, 2019, 11:51 AM Alistair Hey <notifications@github.com> wrote:

> *@Waterdrips* approved this pull request.
>
> LGTM 👍
> Worked a charm on my k3d cluster.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexellis/k3sup/pull/119?email_source=notifications&email_token=ALSRZWPTSDJRDXGKB6ZDKRTQXVF3XA5CNFSM4JWKS5TKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOLHROQ#pullrequestreview-328628410>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALSRZWOMEGIPVMV25QVIQWDQXVF3XANCNFSM4JWKS5TA>
> .
>

alexellis commented on pkg/cmd/minio_app.go:
```diff
@@ -0,0 +1,137 @@
+wget https://dl.min.io/client/mc/release/linux-amd64/mc \
+&& chmod +x mc \
+&& ./mc --help
+
```
Let's just add the `mc config host with above` command?

alexellis replied:
Send in a new PR?

alexellis commented:
Rebase required on this

burtonr commented:
@alexellis I've rebased into a single commit, so this is ready to go now. Thank you

alexellis commented:
Couple of things I noticed:

* Missing the command to add the Minio server after downloading mc, that has been added
* The chart wouldn't deploy on k3s because it was requesting a PV - persistent claim, this is now off by default with an option to enable
* The instructions told the user to download a Linux binary, without looking at the client architecture, that's now dynamic.

> How Has This Been Tested?

Installed Minio app on local microk8s cluster

^ I do wonder if this worked for you? My cluster sat in Pending mode, but maybe microk8s includes a local volume provisioner out of the box?

Thanks for the contribution, hope you'll pick up another one soon 👍 

https://github.com/alexellis/k3sup/commit/9af477914199a92e4daa140bc3dad208745049c4
